### PR TITLE
Fix for documents refCount issue

### PR DESCRIPTION
**Pull request by swmitra**
This PR provides a fix for the document refCount issue reported as part of #12396. This resolves the URL Code hints unit test failures barring the one which is to deal with '@import'.

CC @MarcelGerber @petetnt @nethip @abose for review.

### CODE REVIEW

File: src/document/Document.js [hunk 1]
```diff
@@ -252,6 +250,24 @@ define(function (require, exports, module) {
...
 
+
+    /**
+     * Checks and returns if a full editor exists for the provided pane attached to this document
+     * @param {String} paneId
+     * @return {Editor} Attached editor bound to the provided pane id
+     */
+    Document.prototype._checkAssociatedEditorForPane = function (paneId) {
+        var editorCount, editorForPane;
+        for (editorCount = 0; editorCount < this._associatedFullEditors; editorCount++) {
```
Comment (petetnt) [inline src/document/Document.js]: Missing `length` there: `editorCount < this._associatedFullEditors.length` 🔥

Comment (swmitra) [inline src/document/Document.js]: Sorry. Issue with copy paste from another repo. Fixed now.

**Comment by ficristo** [comment]
I was trying the STR in https://github.com/adobe/brackets/issues/12396, but with this PR when i click on the filename of the inline editor it doesn't open the file.
The cursor is moved at the correct line but in the current file.

**Comment by nethip** [comment]
@swmitra Could you address the query posted by @ficristo ?

**Comment by swmitra** [comment]
@ficristo Thanks a lot for actually trying the fix :+1: 

I need one clarification regarding the behavior you have mentioned. I am guessing that the inline editor and the host editor in your case is pointing to the same document. If that's the case then, even without this fix the behavior is same in master. 

Can you please confirm?

**Comment by ficristo** [comment]
@swmitra I've opened the SpecRunnerUtils.js and put the cursor on 406 (this on current master).
Pressing Ctrl-E it opens the inline editor pointing to EditorManager line 156.
If I click on the name it move the cursor on line 156 but of SpecRunnerUtils.
It should instead open EditorManager at lime 156.
I've tryed this after reload without extensions.

**Comment by swmitra** [comment]
Thanks @ficristo. It's clear now, I will fix it tonight.

**Comment by swmitra** [comment]
@ficristo @MarcelGerber Can you please try with the recent changes? 
Thanks in advance :smile:

**Comment by ficristo** [comment]
Now when I click on the inline editor I have this error:
`/src/editor/EditorManager.js found in pane working set but pane.addView() has not been called for the view created for it`

**Comment by swmitra** [comment]
@ficristo It was getting caused by a missing check in _showEditor(). I had put it back now. URL Code hint test cases are passing as well. Can you please check it once more (hopefully for the last time :smile: ). 
Thanks again for helping us out in verifying the fix :+1:

**Comment by ficristo** [comment]
There is still an error on `Url Code Hints`, not sure if is related to this or not.
I couldn't reproduce the exact steps mentioned by @MarcelGerber in https://github.com/adobe/brackets/issues/12396: I have a different `_refcount` number.
I tested what I could and it seems good; someone with better knowledge should look at it too.

File: src/document/Document.js [hunk 2]
```diff
@@ -241,17 +232,29 @@ define(function (require, exports, module) {
...
 
+
+    /**
+     * Checks and returns if a full editor exists for the provided pane attached to this document
+     * @param {String} paneId
+     * @return {Editor} Attached editor bound to the provided pane id
+     */
+    Document.prototype._checkAssociatedEditorForPane = function (paneId) {
+        var editorCount, editorForPane;
+        for (editorCount = 0; editorCount < this._associatedFullEditors.length; editorCount++) {
```
Comment (nethip) [inline src/document/Document.js]: ++editorCount

**Comment by MarcelGerber** [comment]
@ficristo This one test failure is fixed on master already, so with latest master merged in, all test pass :+1: 
I've followed my repro steps and even though I didn't get the same numbers in the first place, as a result, I had only one document left open. Thumbs up for that! Good work, @swmitra!

**Comment by nethip** [comment]
@swmitra I have taken liberty in changing `editorCount++` to `++editorCount.

**Comment by nethip** [comment]
Since travis is not able to run because of API limitation, I have manually run `grunt build` and the build went fine. So merging this.

Thanks @swmitra 👍 

And thanks to @MarcelGerber @ficristo for helping this PR come to light.

**Comment by swmitra** [comment]
Thanks a lot @nethip @MarcelGerber @ficristo for helping us resolving the issue :+1:

**Pull request merged**